components/scream/src/share/field/field_boundedness_check.hpp

```diff
@@ -0,0 +1,76 @@
+#ifndef SCREAM_FIELD_BOUNDEDNESS_CHECK_HPP
+#define SCREAM_FIELD_BOUNDEDNESS_CHECK_HPP
+
+#include "share/field/field.hpp"
+#include "ekat/util/ekat_math_utils.hpp"
+
+namespace scream
+{
+
+// This field property check returns true if all data in a given field are
+// bounded by the given upper and lower bounds (inclusively), and false if not.
+// It can repair a field that fails the check by clipping all unbounded values
+// to their closest bound.
+template<typename ScalarType, typename Device>
+class FieldBoundednessCheck: public FieldPropertyCheck<ScalarType, Device> {
+public:
+
+  // No default constructor -- we need lower and upper bounds.
+  FieldBoundednessCheck () = delete;
+
+  // Constructor with lower and upper bounds -- can repair fields that fail the check
+  // by overwriting nonpositive values with the given lower bound.
+  explicit FieldBoundednessCheck (ScalarType lower_bound,
+                                  ScalarType upper_bound) :
+    m_lower_bound(lower_bound),
+    m_upper_bound(upper_bound) {
+    EKAT_ASSERT_MSG(lower_bound < upper_bound,
+                    "lower_bound must be less than upper_bound.");
+  }
+
+  // Overrides.
+
+  bool check(const Field<ScalarType, Device>& field) const override {
+    auto host_view = Kokkos::create_mirror_view(field.get_view());
+    Kokkos::deep_copy(host_view, field.get_view());
+    typename Kokkos::MinMax<ScalarType>::value_type minmax;
+    Kokkos::parallel_reduce(host_view.extent(0), KOKKOS_LAMBDA(Int i,
+        typename Kokkos::MinMax<ScalarType>::value_type& mm) {
+      if (i == 0) {
+        mm.min_val = mm.max_val = host_view(0);
+      } else {
+        mm.min_val = ekat::impl::min(mm.min_val, host_view(i));
+        mm.max_val = ekat::impl::max(mm.max_val, host_view(i));
+      }
+    }, Kokkos::MinMax<ScalarType>(minmax));
+    return ((minmax.min_val >= m_lower_bound) && (minmax.max_val <= m_upper_bound));
+  }
+
+  bool can_repair() const override {
+    return true;
```

Comment:
Should we maybe make this a configurable option at construction time? Some atm procs might not be interested in fixing the pb, but just want to error out...

Reply:
Sure, that makes sense.